osbs/tekton.py

```diff
@@ -318,40 +317,45 @@ def log_and_sleep():
+                    if 'object' not in j:
+                        logger.error("Watch event has no 'object': %s", j)
+                        continue
+                    if 'type' not in j:
+                        logger.error("Watch event has no 'type': %s", j)
+                        continue
+
+                    # Avoid races. We've already asked the server to tell us
+                    # about changes to the object, but now ask for a fresh
+                    # copy of the object as well. This is to catch the
+                    # situation where the object changed before the call to
+                    # this method, or in between retries in this method.
+                    logger.debug("retrieving fresh version of object %s", resource_name)
+                    fresh_response = self.get(get_url)
+                    check_response(fresh_response)
+                    yield fresh_response.json()
+
             # we're already retrying, so there's no need to panic just because of a bad response
             except OsbsResponseException as exc:
                 bad_responses += 1
                 if bad_responses > MAX_BAD_RESPONSES:
                     raise exc
-                else:
-                    # check_response() already logged the message, so just report that we're
-                    # sleeping and retry
-                    log_and_sleep()
-                    continue
-
-            for line in response.iter_lines():
-                encoding = guess_json_utf(line)
-                try:
-                    j = json.loads(line.decode(encoding))
-                except ValueError:
-                    logger.error("Cannot decode watch event: %s", line)
-                    continue
-                if 'object' not in j:
-                    logger.error("Watch event has no 'object': %s", j)
-                    continue
-                if 'type' not in j:
-                    logger.error("Watch event has no 'type': %s", j)
-                    continue
-
-                # Avoid races. We've already asked the server to tell us
-                # about changes to the object, but now ask for a fresh
-                # copy of the object as well. This is to catch the
-                # situation where the object changed before the call to
-                # this method, or in between retries in this method.
-                logger.debug("retrieving fresh version of object %s", resource_name)
-                fresh_response = self.get(get_url)
-                check_response(fresh_response)
-                yield fresh_response.json()
+
+            except OsbsException as exc:
+                if (not isinstance(exc.cause, requests.ConnectionError) and
+                        not isinstance(exc.cause, requests.Timeout)):
+                    raise
+            except requests.exceptions.ConnectionError:
+                pass
+            except requests.exceptions.Timeout:
+                pass
```

Comment:
Why do we now also retry on these? Doesn't the underlying `HttpStream` class already do these retries?

Reply:
HttpsStream handles those, but this is also done in _stream_logs

and since we are now using watcher way more than in osbs1, I've added it here as well